pkg/remote/git.go

```diff
@@ -0,0 +1,92 @@
+/*
+   Copyright 2020 Docker Compose CLI authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package remote
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/compose-spec/compose-go/cli"
+	"github.com/compose-spec/compose-go/loader"
+	"github.com/docker/cli/cli/config"
+	"github.com/docker/compose/v2/pkg/api"
+	"github.com/moby/buildkit/util/gitutil"
+	"github.com/opencontainers/go-digest"
+)
+
+func NewGitRemoteLoader() (loader.ResourceLoader, error) {
+	cache := filepath.Join(config.Dir(), "compose", "remotes")
+	err := os.MkdirAll(cache, 0o700)
+	return gitRemoteLoader{
+		cache: cache,
+	}, err
+}
+
+type gitRemoteLoader struct {
+	cache string
+}
+
+func (g gitRemoteLoader) Accept(path string) bool {
+	_, err := gitutil.ParseGitRef(path)
+	return err == nil
+}
+
+func (g gitRemoteLoader) Load(ctx context.Context, path string) (string, error) {
+	ref, err := gitutil.ParseGitRef(path)
+	if err != nil {
+		return "", err
+	}
+	path = fmt.Sprintf("%s#%s", ref.Remote, ref.Commit)
+	local := filepath.Join(g.cache, digest.FromString(path).Hex())
+	if _, err := os.Stat(local); os.IsNotExist(err) {
```

Comment:
Have we thought about the UX for updates here? If my `ref` is a branch (`main`), it's never going to get updated.

`--remote-update-policy=always`? 😅

Reply:
I changed the implementation so it relies on `git ls-remote` to check remote for commit matching branch name, and checkout selected commit if not already known